fixieai/agents-ts/src/embed.ts

```diff
@@ -0,0 +1,73 @@
+import got from 'got';
+
+function isURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+function uriFromBase64(base64Data: string): string {
+  return `data:base64,${base64Data}`;
+}
+
+export interface SerializedEmbed {
+  content_type: string;
+  uri: string;
+}
+
+/**
+ * A binary object attached to a Message.
+ */
+export class Embed {
+  private constructor(
+    /**
+     * The MIME content type of the object, e.g., "image/png" or "application/json".
+     *
+     * https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+     */
+    public readonly contentType: string,
+    /**
+     * The base64-encoded data for this embed.
+     */
+    public readonly base64Data: string,
+    public readonly uri: string,
+  ) {
+  }
+
+  serialize(): SerializedEmbed {
+    return {
+      // This is to match the Python agent's API.
+      // eslint-disable-next-line camelcase
+      content_type: this.contentType,
+      uri: uriFromBase64(this.base64Data),
```

Comment:
This doesn't have to be base64, it could be a typical URL.

Reply:
IOW you should be able to do either do fromBase64("image/png", <base64 data>) or Embed("image/png", "http://image.png") (and ideally with the last one the content type would be optional)

Reply:
Let's chat in person – my understanding from @petersalas was that the func should always emit an embed with a `base64` data URL.

Reply:
Pushed changes after chatting in person.

Reply:
I think we _could_ convert all URIs to data URIs in the agent but it's not necessary to do so. One catch that comes to mind is that it requires the contents to be loaded in memory and might cause OOMs. (Agent OOMs are better than Fixie OOMs, but the latter we can address by being deliberate about how we manage those requests.) So if we want to avoid having a hard limit on embed size we're probably better off passing them indirectly back to the service.